### PR TITLE
Dashboards: Allow updating a dashboard if the user doesn't have access to the parent folder

### LINKED ISSUE #76068: Unable to save a dashboard when lacking permissions on parent folder
When a user has edit permissions on a dashboard, but permissions removed from the folder the dashboard is in, the API fails when attempting to save the dashboard with a 500 error with

```json
{"error":"access denied to folder","message":"Error while checking folder ID","traceID":""}
```

https://github.com/grafana/grafana/assets/46142/7c62bc7c-758d-4bf3-a2b8-e8c9afd13007

### PR DESCRIPTION
**Pull request by IevaVasiljeva**
**What is this feature?**

Refactors the code to allow updating a dashboard if the user doesn't have access to the parent folder:
* don't fetch the folder from the APIs, instead check if the folder is valid from the dashboard service (this seems like a more appropriate place for validation);
* fetch the parent folder using folder store, so we can fetch it even if the caller doesn't have access to the folder.

**Why do we need this feature?**

To allow users to update dashboards if they don't have access to the parent folder.

**Who is this feature for?**

Everyone.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/76068

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/api/dashboard.go [hunk 1]
```diff
@@ -386,25 +385,6 @@ func (hs *HTTPServer) postDashboard(c *contextmodel.ReqContext, cmd dashboards.S
...
-		if err != nil {
-			if errors.Is(err, dashboards.ErrFolderNotFound) {
-				return response.Error(http.StatusBadRequest, "Folder not found", err)
-			}
-			return response.Error(http.StatusInternalServerError, "Error while checking folder ID", err)
-		}
-		// nolint:staticcheck
-		cmd.FolderID = folder.ID
-		cmd.FolderUID = folder.UID
-	}
```
Comment (IevaVasiljeva) [inline pkg/api/dashboard.go:407]: This logic is now carried out [here](https://github.com/grafana/grafana/pull/78075/files#diff-56c23b161208f11acd3ddd098e59329339a548161430a3d7303a1e80e5ed2c8fR134-R148).

File: pkg/services/dashboards/database/database.go [hunk 2]
```diff
@@ -330,19 +330,6 @@ func getExistingDashboardByIDOrUIDForUpdate(sess *db.Session, dash *dashboards.D
...
-		folderExists, err := sess.Where("org_id=? AND id=? AND is_folder=?", dash.OrgID, dash.FolderID,
-			dialect.BooleanStr(true)).Get(&existingFolder)
-		if err != nil {
-			return false, fmt.Errorf("SQL query for folder failed: %w", err)
-		}
-
-		if !folderExists {
-			return false, dashboards.ErrDashboardFolderNotFound
-		}
-	}
```
Comment (IevaVasiljeva) [inline pkg/services/dashboards/database/database.go:344]: This logic is now carried out [here for dashboards](https://github.com/grafana/grafana/pull/78075/files#diff-56c23b161208f11acd3ddd098e59329339a548161430a3d7303a1e80e5ed2c8fR134-R148) and [here for folders](https://github.com/grafana/grafana/pull/78075/files#diff-62d73bdc0f7fc91a4f8dab955eba4b288815692725c72b81c1f36b87c5f123c0R864-R866).

File: pkg/services/folder/folderimpl/folder.go [hunk 3]
```diff
@@ -856,12 +856,14 @@ func (s *Service) buildSaveDashboardCommand(ctx context.Context, dto *dashboards
 		return nil, dashboards.ErrDashboardTitleEmpty
 	}
 
-	if dash.IsFolder && dash.FolderID > 0 {
-		return nil, dashboards.ErrDashboardFolderCannotHaveParent
```
Comment (IevaVasiljeva) [inline pkg/services/folder/folderimpl/folder.go:860]: Cleanup: this is an outdated check - we support nested folders now.

File: pkg/services/folder/folderimpl/folder.go [hunk 4]
```diff
@@ -856,12 +856,14 @@ func (s *Service) buildSaveDashboardCommand(ctx context.Context, dto *dashboards
...
+	if strings.EqualFold(dash.Title, dashboards.RootFolderName) {
+		return nil, dashboards.ErrDashboardFolderNameExists
 	}
 
-	if dash.IsFolder && strings.EqualFold(dash.Title, dashboards.RootFolderName) {
```
Comment (IevaVasiljeva) [inline pkg/services/folder/folderimpl/folder.go:863]: Cleanup: safe to remove `dash.IsFolder`, as this code is a part of the folder service, so we know that `dash.IsFolder` is true.

**Review by Jguer — APPROVED** [review]

**Pull request merged**